core/lib/scmi-server/scmi_clock_consumer.c

```diff
@@ -0,0 +1,278 @@
+	struct clk *clock;
+	uint32_t flags;
+};
+
+/*
+ * struct scmi_clock_consumer: context of scmi_clock_consumer.c for one channel
+ *
+ * @plat_clocks: platform clocks exposed with SCMI
+ * @plat_clocks_flags: capabilities of an plat_clocks
+ * @plat_clocks_count: number of plat_clocks and plat_clocks_flags
+ * @link: link for list of scmi_clock_consumer
+ */
+struct scmi_clock_consumer {
+	struct clk *plat_clocks;
+	uint32_t *plat_clocks_flags;
+	unsigned int plat_clocks_count;
+
+	SLIST_ENTRY(scmi_clock_consumer) link;
+};
+
+static SLIST_HEAD(scmi_clock_consumer_head, scmi_clock_consumer) ctx =
+	SLIST_HEAD_INITIALIZER(&ctx);
+
+static TEE_Result plat_scmi_clk_get_rates_array(struct clk *clk, size_t index,
+						unsigned long *rates,
+						size_t *nb_elts)
+{
+	uint32_t plat_clock_flags = *(uint32_t *)clk->priv;
+
+	if (!nb_elts)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (plat_clock_flags & SCMI_CLOCK_ALLOW_SET_RATE)
+		return clk_get_rates_array(clk->parent, index, rates, nb_elts);
+
+	if (!rates || !*nb_elts) {
+		*nb_elts = 1;
+		return TEE_SUCCESS;
+	}
+
+	if (index)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	assert(rates);
+
+	/*
+	 * Clocks not exposed have no effective parent/platform clock.
+	 * Report a 0 Hz rate in this case.
+	 */
+	if (clk->parent)
+		*rates = clk_get_rate(clk->parent);
+	else
+		*rates = 0;
+
+	*nb_elts = 1;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result plat_scmi_clk_get_rates_steps(struct clk *clk,
+						unsigned long *min,
+						unsigned long *max,
+						unsigned long *step)
+{
+	uint32_t plat_clock_flags = *(uint32_t *)clk->priv;
+
+	if (plat_clock_flags & SCMI_CLOCK_ALLOW_SET_RATE) {
+		*min = 0;
+		*max = UINT32_MAX;
+		*step = 0;
+	} else {
+		*min = clk->rate;
+		*max = *min;
+		*step = 0;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static const struct clk_ops plat_scmi_clk_ops = {
+	.get_rates_array = plat_scmi_clk_get_rates_array,
+	.get_rates_steps = plat_scmi_clk_get_rates_steps,
+};
+
+TEE_Result optee_scmi_server_init_clocks(const void *fdt, int node,
+					 struct scpfw_agent_config *agent_cfg,
+					 struct scpfw_channel_config
+						*channel_cfg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct scmi_clock_consumer *consumer = NULL;
+	struct scmi_server_clock *s_clocks = NULL;
+	size_t s_clocks_count = 0;
+	int subnode = 0, item_node = 0;
+	bool have_subnodes = false;
+	size_t n = 0;
+	size_t i = 0;
+
+	item_node = fdt_subnode_offset(fdt, node, "clocks");
+	if (item_node < 0)
+		return TEE_SUCCESS;
+
+	/* Compute the number of domains to allocate */
+	fdt_for_each_subnode(subnode, fdt, item_node) {
+		paddr_t reg = fdt_reg_base_address(fdt, subnode);
+
+		assert(reg != DT_INFO_INVALID_REG);
+		if ((size_t)reg > s_clocks_count)
+			s_clocks_count = (uint32_t)reg;
+
+		have_subnodes = true;
+	}
+
+	if (!have_subnodes)
+		return TEE_SUCCESS;
+
+	consumer = calloc(1, sizeof(*consumer));
+	if (!consumer)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	SLIST_INSERT_HEAD(&ctx, consumer, link);
```

Comment:
Could insert in the list at function end, upon its success. It would prevent the need to remove the element in the few error cases in this function.